lib/3scale/backend/user.rb

```diff
@@ -77,15 +77,32 @@ def self.save!(attributes)
       def self.delete!(service_id, username)
         service = Service.load_by_id(service_id)
         raise UserRequiresValidService if service.nil?
-        ServiceUserManagementUseCase.new(service, username).delete
-        clear_cache(service_id, username)
-        storage.del(self.key(service_id, username))
+        delete_users_in_batch(service_id, [username])
+      end
+
+      def self.delete_all(service_id)
+        service = Service.load_by_id(service_id)
+        raise UserRequiresValidService if service.nil?
+        current_scan_cursor_position = "0"
+        finished = false
+        users_set_key = service_users_set_key(service_id)
+        while !finished
+          (current_scan_cursor_position, current_usernames) =
+            storage.sscan(users_set_key, current_scan_cursor_position,
```

Comment:
Have you considered using `scard` and `spop` operations to delete all elements of the set? At least seems less risky than using cursors on potentially big data sets.

Reply:
When this was implemented, we were using Redis server 2.8 in SaaS. The spop command does not support specifying the number of elements to delete until Redis 3.2 (https://redis.io/commands/spop) so we could not use this command and that was the reason we opted to use cursor-based deletion.
However, recently we have upgraded our Redis server version so we could replace the cursor to use scard and spop.
If we do this we would need to update the apisonator development Docker images to use a newer version of Redis too because right now it is still using 2.8 and would stop working.
Right now we have the following Redis versions:
SaaS: Redis 4.X
On-Prem: Redis 3.2

If we upgrade Redis server version in the development images I'd say to use Redis 3.2 (the lower version). This would mean that we would not be using in dev the same version that in SaaS but we would be testing a version where all commands should be compatible in both versions. What do you think?

Reply:
@miguelsorianod I think we should update the development images as you suggested so we can change this to use spop :+1:

Reply:
:+1: . I've done this in https://github.com/3scale/apisonator/pull/69

Reply:
I upgraded the dev images and we are now using Redis 3.2.X for our dev images. However, the Redis ruby client version that we use in some of our services is 3.2.2, but the support for spop with the number of elements to pop was added to 3.3.2. At this moment we cannot upgrade to that version in those services due to a bug we found with reconnections https://github.com/redis/redis-rb/issues/668. We'll continue investigating on that and for the moment to be able to proceed we'll leave this with scan.